ee/clickhouse/queries/related_actors_query.py

```diff
@@ -0,0 +1,125 @@
+
+from django.utils.timezone import now
+
+from ee.clickhouse.client import sync_execute
+from ee.clickhouse.sql.person import GET_TEAM_PERSON_DISTINCT_IDS
+from posthog.models.filters.utils import validate_group_type_index
+from posthog.models.group_type_mapping import GroupTypeMapping
+from posthog.models.property import GroupTypeIndex
+
+DISTINCT_IDS_JOIN = f"JOIN ({GET_TEAM_PERSON_DISTINCT_IDS}) pdi on e.distinct_id = pdi.distinct_id"
+
+
+class RelatedActorsResponse(TypedDict):
+    type: Literal["person", "group"]
+    group_type_index: Optional[GroupTypeIndex]
+    id: str
+    person: Optional[Dict]
+
+
+class RelatedActorsQuery:
+    """
+    This query calculates other groups and persons that are related to a person or a group.
+
+    Two actors are considered related if they have had shared events in the past 90 days.
+    """
+
+    def __init__(self, team_id: int, group_type_index: Optional[Union[GroupTypeIndex, str]], id: str):
+        self.team_id = team_id
+        self.group_type_index = validate_group_type_index("group_type_index", group_type_index)
+        self.id = id
+
+    def run(self) -> List[RelatedActorsResponse]:
+        results = self._query_related_people()
+        for group_type_mapping in GroupTypeMapping.objects.filter(team_id=self.team_id):
+            results.extend(self._query_related_groups(group_type_mapping.group_type_index))
+        return results
+
+    @property
+    def is_aggregating_by_groups(self) -> bool:
+        return self.group_type_index is not None
+
+    def _query_related_people(self) -> List[RelatedActorsResponse]:
+        if not self.is_aggregating_by_groups:
+            return []
+
+        # :KLUDGE: We need to fetch distinct_id + person properties to be able to link to user properly.
+        rows = sync_execute(
+            f"""
+            SELECT person_id, any(e.distinct_id), any(person_props)
+            FROM events e
+            {DISTINCT_IDS_JOIN}
+            JOIN (
+                SELECT id, any(properties) as person_props
+                FROM person
+                WHERE team_id = %(team_id)s
+                GROUP BY id
+                HAVING max(is_deleted) = 0
+            ) person ON pdi.person_id = person.id
+            WHERE team_id = %(team_id)s
+              AND timestamp > %(after)s
+              AND timestamp < %(before)s
+              AND {self._filter_clause}
+            GROUP BY person_id
+            """,
+            self._params,
+        )
+
+        return [
+            RelatedActorsResponse(
+                type="person",
+                group_type_index=None,
+                id=person_id,
+                person={"distinct_ids": [distinct_id], "properties": json.loads(person_props)},
+            )
+            for (person_id, distinct_id, person_props) in rows
+        ]
+
+    def _query_related_groups(self, group_type_index: GroupTypeIndex) -> List[RelatedActorsResponse]:
+        if group_type_index == self.group_type_index:
```

Comment:
confusing, does this mean you're looking for related groups that aren't the same type?

Reply:
Exactly. Related actors are other groups and persons which share events with the actor we're querying from

Reply:
Not just other groups, but groups from other group types. I think worth clarifying in the UI?

Oh, I just realised, that groups of the same type can't ever have shared events anyway - since one event -> one group (of the same type). D'oh. That makes sense.

---

Thinking out loud about the clarifying UI bit:

Say `$group_0` is `Companies`, and I have two groups: `Tesla` and `SpaceX`. When I'm looking at `Tesla`, if I raise the question: Why don't I see `SpaceX` in the related groups? implies we need a UI clarification about what related groups mean.

That sounds like a valid question someone might raise. However, okay with pushing it to later, when people do raise it.

Reply:
Yep, this is why I added the tooltip but we can always iterate the wording later.